test/src/unit-diagnostics.cpp

```diff
@@ -184,4 +185,21 @@ TEST_CASE("Better diagnostics")
         j["second"] = value;
         j2["something"] = j;
     }
+
+    SECTION("Regression test for issue #3007 - Parent pointers properly set when using update()")
+    {
+        json j = json::object();
+
+        {
+            json j2 = json::object();
+            j2["one"] = 1;
+
+            j.update(j2);
+        }
+
+        for (auto const& kv : j)
+        {
+            CHECK(kv.m_parent == &j);
```

Comment:
I do not like this test too much as it is using private members. And since the tests are executed in debug mode, the invariant is checked which would trigger an assertion in case of an error.

Instead, you could do the following (not tested):

```cpp
j["one"].at(0);
```

This should trigger an exception, because `j["one"]` is a number and `at` is only defined for `null` or arrays. This exception should then contain a proper JSON path which shows that parents are properly set. This can be checked with `CHECK_THROWS_WITH_AS`.

Reply:
Calling `update()` does not trigger an assertion when checking the invariant on the newly added members. I.e. this test properly fails when it is ran without the `m_parent` fix. (Or I'm misunderstanding what you're saying.)

I do agree that using private members for the test is not great though. So I've modified it to check the message thrown by `at()` as you suggested. I've also the fix to the 2nd `update()` overload which I had overlooked (test added as well).

Reply:
Thanks for checking! If the assertion is not triggered, then I was wrong in my assumption. Running the test on the `develop` branch does not trigger the assertion, but the raised exception

```
[json.exception.type_error.304] cannot use at() with number
```

does not contain a JSON Pointer. So the unfixed branch would at least fail this test.

Furthermore, you could add the original code of #3007 

```cpp
  nlohmann::json root = nlohmann::json::array();
  nlohmann::json lower = nlohmann::json::object();

  {
    nlohmann::json lowest = nlohmann::json::object();
    lowest["one"] = 1;
    
    lower.update(lowest);
  }

  root.push_back(lower);
```

as a regression test.

Reply:
> ```
> [json.exception.type_error.304] cannot use at() with number
> ```
> 
> does not contain a JSON Pointer. So the unfixed branch would at least fail this test.

Yes, that's what I'm currently testing for.

> Furthermore, you could add the original code of #3007 as a regression test.

I could change the current test, but then the new test (i.e. my example cod for #3007) would fire an assertion without the fix. So the regression would not run any other tests.

Do you prefer that over the current test implementation? If not, I guess this can be merged?

Reply:
I did not mean changing one of your added test, but *adding* the code of #3007 as new test. This test should not trigger in assertion in your branch, right?

Reply:
That's true. I've added the code and just pushed it.